stan/math/prim/mat/functor/mpi_parallel_call.hpp

```diff
@@ -0,0 +1,488 @@
+
+    // check that cached sizes are the same as just collected from
+    // this evaluation
+    for (int i = 0; i < num_local_jobs; ++i) {
+      if (world_f_out[start_job + i] != local_f_out[i]) {
+        local_ok = 0;
+        break;
+      }
+    }
+
+    const std::size_t size_world_f_out = sum(world_f_out);
+    matrix_d world_result(num_outputs_per_job_, size_world_f_out);
+
+    std::vector<int> chunks_result(world_size_, 0);
+    for (std::size_t i = 0, ij = 0; i != world_size_; ++i)
+      for (int j = 0; j != job_chunks[i]; ++j, ++ij)
+        chunks_result[i] += world_f_out[ij] * num_outputs_per_job_;
+
+    // collect results on root
+    boost::mpi::gatherv(world_, local_output.data(), chunks_result[rank_],
+                        world_result.data(), chunks_result, 0);
+
+    // let root know if all went fine everywhere
+    int cluster_status = 0;
+    boost::mpi::reduce(world_, local_ok, cluster_status, std::plus<int>(), 0);
+    bool all_ok = cluster_status == static_cast<int>(world_size_);
+
+    // on the workers all is done now.
+    if (rank_ != 0)
+      return result_t();
+
+    // in case something went wrong we throw on the root
+    if (cluster_status != static_cast<int>(world_size_))
+      throw std::domain_error("Error during MPI evaluation.");
+
+    return combine_(world_result, world_f_out);
+  }
+
+ private:
+  template <typename T_cache>
+  void scatter_array_2d_cached(typename T_cache::cache_t& data) {
+    // distribute data only if not in cache yet
+    if (T_cache::is_valid()) {
+      return;
+    }
+
+    // number of elements of each data item must be transferred to
+    // the workers
+    std::vector<int> data_dims = dims(data);
+    data_dims.resize(2);
+
+    boost::mpi::broadcast(world_, data_dims.data(), 2, 0);
+
+    const std::vector<int> job_chunks = mpi_map_chunks(data_dims[0], 1);
+    const std::vector<int> data_chunks
+        = mpi_map_chunks(data_dims[0], data_dims[1]);
+
+    auto flat_data = to_array_1d(data);
+    decltype(flat_data) local_flat_data(data_chunks[rank_]);
+
+    boost::mpi::scatterv(world_, flat_data.data(), data_chunks,
+                         local_flat_data.data(), 0);
+
+    std::vector<decltype(flat_data)> local_data;
+    auto local_iter = local_flat_data.begin();
+    for (int i = 0; i != job_chunks[rank_]; ++i) {
+      typename T_cache::cache_t::value_type const data_elem(
+          local_iter, local_iter + data_dims[1]);
+      local_data.push_back(data_elem);
+      local_iter += data_dims[1];
+    }
+
+    // finally we cache it locally
+    T_cache::store(local_data);
+  }
+
+  template <typename T_cache>
+  void broadcast_1d_cached(typename T_cache::cache_t& data) {
+    if (T_cache::is_valid()) {
+      return;
+    }
+
+    std::size_t data_size = data.size();
+    boost::mpi::broadcast(world_, data_size, 0);
+
+    typename std::remove_cv<typename T_cache::cache_t>::type local_data = data;
```

Comment:
This is a little too fancy for me with the removal of qualifiers.  Can you just declare this locally?

Also, rather than assigning to `data` then resizing, it'd be better just to use iterators to set it to the right things to begin with if it's a `std::vector` under all that typing.

Reply:
I can check again, but I think I need the `remove_cv` here. The complication here is that the same code does slightly different things when run on the root and on a worker. The root sends while the workers receive the data (using the same code chunk).